### ISSUE TITLE
ConcurrentModificationException in OpenRewrite recipe

### ISSUE THREAD

**Issue by sk-br** [aws/aws-sdk-java-v2#5666]
### Describe the bug

The (dry-)run of the OpenRewrite recipe crashes at our multi module maven project.

### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Expected Behavior

The (dry-)run of the OpenRewrite recipe should terminate successfully.

### Current Behavior

It breaks throwing a ConcurrentModificationException at
`software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisitSourceFile(ChangeSdkType.java:208)`

### Reproduction Steps

Run the script on a multi module project, where multiple modules use the AWS SDK.

### Possible Solution

Use a `ConcurrentHashMap` instead of a normal `HashMap` - see https://github.com/aws/aws-sdk-java-v2/pull/5665

### Additional Information/Context

Maven version: 3.9.6 (regular maven, not mvnd)

### AWS Java SDK version used

2.28.19-PREVIEW

### JDK version used

OpenJDK Runtime Environment Temurin-17.0.10+7 (build 17.0.10+7)

### Operating System and version

MacOS 14.6.1

**Comment by debora-ito** [comment]
@sk-br can you share the full stacktrace, please?

**Comment by sk-br** [comment]
Here your are:
```
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.openrewrite.maven:rewrite-maven-plugin:5.42.2:dryRun (default-cli) on project mir-frontend-ng: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:5.42.2:dryRun failed: Error while visiting mir-domain-jpa/src/main/java/de/br/mir/storage/S3StorageClient.java: java.util.ConcurrentModificationException: null
  java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1597)
  java.base/java.util.HashMap$ValueIterator.next(HashMap.java:1625)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisitSourceFile(ChangeSdkType.java:208)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:199)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:83)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:251)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:150)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$6(RecipeRunCycle.java:182)
  io.micrometer.core.instrument.AbstractTimer.recordCallable(AbstractTimer.java:178)
  org.openrewrite.table.RecipeRunStats.recordEdit(RecipeRunStats.java:67)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$7(RecipeRunCycle.java:178)
  org.openrewrite.scheduling.RecipeStack.reduce(RecipeStack.java:57)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$8(RecipeRunCycle.java:151)
  org.openrewrite.internal.InMemoryLargeSourceSet.lambda$edit$0(InMemoryLargeSourceSet.java:66)
  org.openrewrite.internal.ListUtils.map(ListUtils.java:176)
  org.openrewrite.internal.InMemoryLargeSourceSet.edit(InMemoryLargeSourceSet.java:65)
  ...
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:333)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:283)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:226)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:407)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:348)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:5.42.2:dryRun failed: Error while visiting mir-domain-jpa/src/main/java/de/br/mir/storage/S3StorageClient.java: java.util.ConcurrentModificationException: null
  java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1597)
  java.base/java.util.HashMap$ValueIterator.next(HashMap.java:1625)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisitSourceFile(ChangeSdkType.java:208)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:199)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:83)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:251)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:150)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$6(RecipeRunCycle.java:182)
  io.micrometer.core.instrument.AbstractTimer.recordCallable(AbstractTimer.java:178)
  org.openrewrite.table.RecipeRunStats.recordEdit(RecipeRunStats.java:67)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$7(RecipeRunCycle.java:178)
  org.openrewrite.scheduling.RecipeStack.reduce(RecipeStack.java:57)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$8(RecipeRunCycle.java:151)
  org.openrewrite.internal.InMemoryLargeSourceSet.lambda$edit$0(InMemoryLargeSourceSet.java:66)
  org.openrewrite.internal.ListUtils.map(ListUtils.java:176)
  org.openrewrite.internal.InMemoryLargeSourceSet.edit(InMemoryLargeSourceSet.java:65)
  ...
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:133)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:283)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:226)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:407)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:348)
Caused by: java.lang.RuntimeException: Error while visiting mir-domain-jpa/src/main/java/de/br/mir/storage/S3StorageClient.java: java.util.ConcurrentModificationException: null
  java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1597)
  java.base/java.util.HashMap$ValueIterator.next(HashMap.java:1625)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisitSourceFile(ChangeSdkType.java:208)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:199)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:83)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:251)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:150)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$6(RecipeRunCycle.java:182)
  io.micrometer.core.instrument.AbstractTimer.recordCallable(AbstractTimer.java:178)
  org.openrewrite.table.RecipeRunStats.recordEdit(RecipeRunStats.java:67)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$7(RecipeRunCycle.java:178)
  org.openrewrite.scheduling.RecipeStack.reduce(RecipeStack.java:57)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$8(RecipeRunCycle.java:151)
  org.openrewrite.internal.InMemoryLargeSourceSet.lambda$edit$0(InMemoryLargeSourceSet.java:66)
  org.openrewrite.internal.ListUtils.map(ListUtils.java:176)
  org.openrewrite.internal.InMemoryLargeSourceSet.edit(InMemoryLargeSourceSet.java:65)
  ...
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer$1.lambda$preVisit$0 (AbstractRewriteBaseRunMojo.java:380)
    at java.util.Optional.ifPresent (Optional.java:178)
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer$1.preVisit (AbstractRewriteBaseRunMojo.java:377)
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer$1.preVisit (AbstractRewriteBaseRunMojo.java:373)
    at org.openrewrite.TreeVisitor.visit (TreeVisitor.java:245)
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer.getRecipeErrors (AbstractRewriteBaseRunMojo.java:384)
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer.getFirstException (AbstractRewriteBaseRunMojo.java:364)
    at org.openrewrite.maven.AbstractRewriteDryRunMojo.execute (AbstractRewriteDryRunMojo.java:72)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:283)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:226)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:407)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:348)
```

**Comment by debora-ito** [comment]
@sk-br Thank you for providing all the info so far. 

I'm sorry, we have one more question: what's the actual mvn command you used? We are interested in the options used in the command.

**Comment by sk-br** [comment]
I used the following:
```
sdkversion="2.28.19-PREVIEW"

mvn -e org.openrewrite.maven:rewrite-maven-plugin:dryRun \
  -Drewrite.recipeArtifactCoordinates=software.amazon.awssdk:v2-migration:${sdkversion} \
  -Drewrite.activeRecipes=software.amazon.awssdk.v2migration.AwsSdkJavaV1ToV2
```

**Comment by SamBarker** [comment]
I've got very similar looking stack trace running
```
mvn -e -Pfast -Pjava17 -Pskip-webui-build -pl :flink-s3-fs-base -am -amd org.openrewrite.maven:rewrite-maven-plugin:dryRun \                                                                             
  -Drewrite.recipeArtifactCoordinates=software.amazon.awssdk:v2-migration:2.28.29-PREVIEW \ 
-Drewrite.activeRecipes=software.amazon.awssdk.v2migration.AwsSdkJavaV1ToV2
```
On https://github.com/apache/flink 

```
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.openrewrite.maven:rewrite-maven-plugin:5.43.0:dryRun (default-cli) on project flink-failure-enricher-test: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:5.43.0:dryRun failed: Error while visiting flink-filesystems/flink-s3-fs-base/src/main/java/com/amazonaws/services/s3/model/transform/XmlResponsesSaxParser.java: java.util.ConcurrentModificationException: null
  java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1511)
  java.base/java.util.HashMap$ValueIterator.next(HashMap.java:1539)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisitSourceFile(ChangeSdkType.java:208)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:199)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:83)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:253)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:157)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$6(RecipeRunCycle.java:186)
  io.micrometer.core.instrument.AbstractTimer.recordCallable(AbstractTimer.java:178)
  org.openrewrite.table.RecipeRunStats.recordEdit(RecipeRunStats.java:67)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$7(RecipeRunCycle.java:182)
  org.openrewrite.scheduling.RecipeStack.reduce(RecipeStack.java:57)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$8(RecipeRunCycle.java:155)
  org.openrewrite.internal.InMemoryLargeSourceSet.lambda$edit$0(InMemoryLargeSourceSet.java:66)
  org.openrewrite.internal.ListUtils.map(ListUtils.java:177)
  org.openrewrite.internal.InMemoryLargeSourceSet.edit(InMemoryLargeSourceSet.java:65)
  ...
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:333)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:283)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:226)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:407)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:348)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:5.43.0:dryRun failed: Error while visiting flink-filesystems/flink-s3-fs-base/src/main/java/com/amazonaws/services/s3/model/transform/XmlResponsesSaxParser.java: java.util.ConcurrentModificationException: null
  java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1511)
  java.base/java.util.HashMap$ValueIterator.next(HashMap.java:1539)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisitSourceFile(ChangeSdkType.java:208)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:199)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:83)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:253)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:157)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$6(RecipeRunCycle.java:186)
  io.micrometer.core.instrument.AbstractTimer.recordCallable(AbstractTimer.java:178)
  org.openrewrite.table.RecipeRunStats.recordEdit(RecipeRunStats.java:67)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$7(RecipeRunCycle.java:182)
  org.openrewrite.scheduling.RecipeStack.reduce(RecipeStack.java:57)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$8(RecipeRunCycle.java:155)
  org.openrewrite.internal.InMemoryLargeSourceSet.lambda$edit$0(InMemoryLargeSourceSet.java:66)
  org.openrewrite.internal.ListUtils.map(ListUtils.java:177)
  org.openrewrite.internal.InMemoryLargeSourceSet.edit(InMemoryLargeSourceSet.java:65)
  ...
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:133)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:283)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:226)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:407)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:348)
Caused by: java.lang.RuntimeException: Error while visiting flink-filesystems/flink-s3-fs-base/src/main/java/com/amazonaws/services/s3/model/transform/XmlResponsesSaxParser.java: java.util.ConcurrentModificationException: null
  java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1511)
  java.base/java.util.HashMap$ValueIterator.next(HashMap.java:1539)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisitSourceFile(ChangeSdkType.java:208)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:199)
  software.amazon.awssdk.v2migration.ChangeSdkType$ChangeTypeVisitor.postVisit(ChangeSdkType.java:83)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:253)
  org.openrewrite.TreeVisitor.visit(TreeVisitor.java:157)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$6(RecipeRunCycle.java:186)
  io.micrometer.core.instrument.AbstractTimer.recordCallable(AbstractTimer.java:178)
  org.openrewrite.table.RecipeRunStats.recordEdit(RecipeRunStats.java:67)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$7(RecipeRunCycle.java:182)
  org.openrewrite.scheduling.RecipeStack.reduce(RecipeStack.java:57)
  org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$8(RecipeRunCycle.java:155)
  org.openrewrite.internal.InMemoryLargeSourceSet.lambda$edit$0(InMemoryLargeSourceSet.java:66)
  org.openrewrite.internal.ListUtils.map(ListUtils.java:177)
  org.openrewrite.internal.InMemoryLargeSourceSet.edit(InMemoryLargeSourceSet.java:65)
  ...
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer$1.lambda$preVisit$0 (AbstractRewriteBaseRunMojo.java:380)
    at java.util.Optional.ifPresent (Optional.java:183)
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer$1.preVisit (AbstractRewriteBaseRunMojo.java:377)
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer$1.preVisit (AbstractRewriteBaseRunMojo.java:373)
    at org.openrewrite.TreeVisitor.visit (TreeVisitor.java:247)
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer.getRecipeErrors (AbstractRewriteBaseRunMojo.java:384)
    at org.openrewrite.maven.AbstractRewriteBaseRunMojo$ResultsContainer.getFirstException (AbstractRewriteBaseRunMojo.java:364)
    at org.openrewrite.maven.AbstractRewriteDryRunMojo.execute (AbstractRewriteDryRunMojo.java:72)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:283)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:226)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:407)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:348)
```

**Comment by debora-ito** [comment]
PR #5665 was merged and released in Java SDK version `2.29.21`.

Please test it out and let us know if the issue persists.

**Comment by SamBarker** [comment]
I can confirm `2.29.21` does fix the error from flink
```
mvn -e -Pfast -Pjava17 -Pskip-webui-build -pl :flink-s3-fs-base -am org.openrewrite.maven:rewrite-maven-plugin:dryRun \ 
-Drewrite.recipeArtifactCoordinates=software.amazon.awssdk:v2-migration:2.29.21-PREVIEW \
  -Drewrite.activeRecipes=software.amazon.awssdk.v2migration.AwsSdkJavaV1ToV2
  ```
  
Generates a patch file.

Thanks.
